cpp/src/arrow/compute/kernel.cc

```diff
@@ -400,9 +399,8 @@ bool InputType::Equals(const InputType& other) const {
       return type_->Equals(*other.type_);
     case InputType::USE_TYPE_MATCHER:
       return type_matcher_->Equals(*other.type_matcher_);
-    default:
-      return false;
   }
+  return false;
```

Comment:
Can we add a comment about "must not be reached" or something here?

Reply:
It might be reached if `kind_` is bogus memory. I could remove this line altogether and let poison (true/false randomly) be returned.

Reply:
I'm assuming @wesm was intentional in having the `default` case here to protect against that kind of UB.

I checked this on godbolt: https://godbolt.org/z/qEaGMGdzh